html2asketch/helpers/convertAngleToFromAndTo.js

```diff
@@ -1,39 +1,125 @@
-  const to = {x: 0.5, y: 1};
-
-  // Learn math or find someone smarter to figure this out correctly
-  switch (angle) {
-    case 'to top':
-    case '360deg':
-    case '0deg':
-      from.y = 1;
-      to.y = 0;
-      break;
-    case 'to right':
-    case '90deg':
-      from.x = 0;
-      from.y = 0.5;
-      to.x = 1;
-      to.y = 0.5;
-      break;
-    case 'to left':
-    case '270deg':
-      from.x = 1;
-      from.y = 0.5;
-      to.x = 0;
-      to.y = 0.5;
-      break;
-    case 'to bottom':
-    case '180deg':
-    default:
-      break;
+export default function convertAngleToFromAndTo(angle, width, height) {
+  function parseAngleToRadians(angle) {
+    const match = angle.match(/^(-?(?:\d+)?(?:\.\d+)?)(deg|grad|rad|turn)$/);
+
+    if (!match) {
+      return Number.NaN;
+    }
+
+    const value = parseFloat(match[1]) + 0;
+    const unit = match[2];
+
+    switch (unit) {
+      case 'deg':
+        return value * (Math.PI / 180);
+      case 'grad':
+        return value * (Math.PI / 200);
+      case 'rad':
+        return value;
+      case 'turn':
+        return value * 2 * Math.PI;
+      default:
+        return Number.NaN;
+    }
   }
 
-  return {
-    from,
-    to,
-  };
+  function fixFloat(float) {
+    const precistion = 7;
+
+    // round with precision and convert -0 to +0
+    return Number.parseFloat(float.toFixed(precistion)) + 0;
+  }
+
+  function calculateFromAndTo({angleInRadians, width, height}) {
+    const angle180degInRadians = parseAngleToRadians('180deg');
+
+    // calculate gradient height
+    const gradientHeight = Math.abs(width * Math.sin(angleInRadians)) + Math.abs(height * Math.cos(angleInRadians));
+
+    // calculate to which is half of gradient times cos/sin of angle
+    const toX = fixFloat((gradientHeight / 2) * Math.sin(angleInRadians));
+    const toY = fixFloat((gradientHeight / 2) * Math.cos(angleInRadians));
+
+    // calculate from which is half of gradient times cos/sin of angle + 180deg
+    const fromX = fixFloat((gradientHeight / 2) * Math.sin(angleInRadians + angle180degInRadians));
+    const fromY = fixFloat((gradientHeight / 2) * Math.cos(angleInRadians + angle180degInRadians));
+
+    return {from: {x: fromX, y: fromY}, to: {x: toX, y: toY}};
+  }
+
+  function normalizeDimensionsForSketch({from, to, width, height}) {
+    const response = {from: {...from}, to: {...to}};
+
+    // y axis should be proportional to height
+    response.from.y /= height;
+    response.to.y /= height;
+
+    // x axis should be proportional to width
+    response.from.x /= width;
+    response.to.x /= width;
+
+    // y axis is reflected downwards in sketch
+    response.from.y *= -1;
+    response.to.y *= -1;
+
+    // let's fixFloat
+    response.from.x = fixFloat(response.from.x);
+    response.from.y = fixFloat(response.from.y);
+    response.to.x = fixFloat(response.to.x);
+    response.to.y = fixFloat(response.to.y);
+
+    // (0,0) is in (0.5,0.5), we need to move whole axis
```

Comment:
small correction: 0.5, 0.5 is origin for the gradient line, as we wanted gradient line to always go through the middle of shape.

Reply:
true. I've corrented the comment